reverted:
--- /var/tmp/portage/sys-kernel/gentoo-sources-4.0.5/work/linux-4.0.5-gentoo/drivers/acpi/ec.c	2015-04-12 17:12:50.000000000 -0500
+++ /usr/src/linux/drivers/acpi/ec.c	2015-06-29 15:36:08.526477654 -0500
@@ -1099,6 +1099,8 @@
 
 	ret = ec_install_handlers(ec);
 
+       acpi_walk_dep_device_list(device->handle);
+
 	/* EC is fully operational, allow queries */
 	clear_bit(EC_FLAGS_QUERY_PENDING, &ec->flags);
 
reverted:
--- /var/tmp/portage/sys-kernel/gentoo-sources-4.0.5/work/linux-4.0.5-gentoo/drivers/acpi/sleep.c	2015-04-12 17:12:50.000000000 -0500
+++ /usr/src/linux/drivers/acpi/sleep.c	2015-06-29 15:36:22.384477334 -0500
@@ -629,6 +629,7 @@
 
 static int acpi_freeze_prepare(void)
 {
+	acpi_enable_wakeup_devices(ACPI_STATE_S0);
 	acpi_enable_all_wakeup_gpes();
 	acpi_os_wait_events_complete();
 	enable_irq_wake(acpi_gbl_FADT.sci_interrupt);
reverted:
--- /var/tmp/portage/sys-kernel/gentoo-sources-4.0.5/work/linux-4.0.5-gentoo/drivers/hid/hid-core.c	2015-04-12 17:12:50.000000000 -0500
+++ /usr/src/linux/drivers/hid/hid-core.c	2015-06-30 18:43:38.987186057 -0500
@@ -704,11 +704,11 @@
 	    type == HID_COLLECTION_PHYSICAL)
 		hid->group = HID_GROUP_SENSOR_HUB;
 
+//	if (hid->vendor == USB_VENDOR_ID_MICROSOFT &&
+//	    (hid->product == USB_DEVICE_ID_MS_TYPE_COVER_3 ||
+//	     hid->product == USB_DEVICE_ID_MS_TYPE_COVER_3_JP) &&
+//	    hid->group == HID_GROUP_MULTITOUCH)
+//		hid->group = HID_GROUP_GENERIC;
-	if (hid->vendor == USB_VENDOR_ID_MICROSOFT &&
-	    (hid->product == USB_DEVICE_ID_MS_TYPE_COVER_3 ||
-	     hid->product == USB_DEVICE_ID_MS_TYPE_COVER_3_JP) &&
-	    hid->group == HID_GROUP_MULTITOUCH)
-		hid->group = HID_GROUP_GENERIC;
 
 	if ((parser->global.usage_page << 16) == HID_UP_GENDESK)
 		for (i = 0; i < parser->local.usage_index; i++)
@@ -1878,8 +1878,8 @@
 	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_DIGITAL_MEDIA_3K) },
 	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_WIRELESS_OPTICAL_DESKTOP_3_0) },
 	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_OFFICE_KB) },
+//	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_TYPE_COVER_3) },
+//	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_TYPE_COVER_3_JP) },
-	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_TYPE_COVER_3) },
-	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_TYPE_COVER_3_JP) },
 	{ HID_USB_DEVICE(USB_VENDOR_ID_MONTEREY, USB_DEVICE_ID_GENIUS_KB29E) },
 	{ HID_USB_DEVICE(USB_VENDOR_ID_MSI, USB_DEVICE_ID_MSI_GT683R_LED_PANEL) },
 	{ HID_USB_DEVICE(USB_VENDOR_ID_NTRIG, USB_DEVICE_ID_NTRIG_TOUCH_SCREEN) },
reverted:
--- /var/tmp/portage/sys-kernel/gentoo-sources-4.0.5/work/linux-4.0.5-gentoo/drivers/hid/hid-microsoft.c	2015-04-12 17:12:50.000000000 -0500
+++ /usr/src/linux/drivers/hid/hid-microsoft.c	2015-06-30 18:44:13.228187334 -0500
@@ -276,10 +276,10 @@
 		.driver_data = MS_NOGET },
 	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_COMFORT_MOUSE_4500),
 		.driver_data = MS_DUPLICATE_USAGES },
+	//{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_TYPE_COVER_3),
+//		.driver_data = MS_HIDINPUT },
+//	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_TYPE_COVER_3_JP),
+//		.driver_data = MS_HIDINPUT },
-	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_TYPE_COVER_3),
-		.driver_data = MS_HIDINPUT },
-	{ HID_USB_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_TYPE_COVER_3_JP),
-		.driver_data = MS_HIDINPUT },
 
 	{ HID_BLUETOOTH_DEVICE(USB_VENDOR_ID_MICROSOFT, USB_DEVICE_ID_MS_PRESENTER_8K_BT),
 		.driver_data = MS_PRESENTER },
reverted:
--- /var/tmp/portage/sys-kernel/gentoo-sources-4.0.5/work/linux-4.0.5-gentoo/drivers/hid/hid-multitouch.c	2015-04-12 17:12:50.000000000 -0500
+++ /usr/src/linux/drivers/hid/hid-multitouch.c	2015-06-29 15:36:37.166476993 -0500
@@ -1246,6 +1246,11 @@
 		MT_USB_DEVICE(USB_VENDOR_ID_TURBOX,
 			USB_DEVICE_ID_TURBOX_TOUCHSCREEN_MOSART) },
 
+	/* Microsoft Type Cover */
+	{ .driver_data = MT_CLS_EXPORT_ALL_INPUTS,
+		MT_USB_DEVICE(USB_VENDOR_ID_MICROSOFT,
+			USB_DEVICE_ID_MS_TYPE_COVER_3) },
+
 	/* Panasonic panels */
 	{ .driver_data = MT_CLS_PANASONIC,
 		MT_USB_DEVICE(USB_VENDOR_ID_PANASONIC,
reverted:
--- /var/tmp/portage/sys-kernel/gentoo-sources-4.0.5/work/linux-4.0.5-gentoo/drivers/media/usb/uvc/uvc_driver.c	2015-04-12 17:12:50.000000000 -0500
+++ /usr/src/linux/drivers/media/usb/uvc/uvc_driver.c	2015-06-29 15:36:31.946477114 -0500
@@ -2379,6 +2379,22 @@
 	  .bInterfaceSubClass	= 1,
 	  .bInterfaceProtocol	= 0,
 	  .driver_info		= UVC_QUIRK_PROBE_MINMAX },
+	/* Microsoft Surface Pro 3 Front */
+	{ .match_flags          = USB_DEVICE_ID_MATCH_DEVICE
+				| USB_DEVICE_ID_MATCH_INT_INFO,
+	  .idVendor             = 0x045e,
+	  .idProduct            = 0x07be,
+	  .bInterfaceClass      = USB_CLASS_VIDEO,
+	  .bInterfaceSubClass   = 1,
+	  .bInterfaceProtocol   = 1 },
+	/* Microsoft Surface Pro 3 Rear */
+	{ .match_flags          = USB_DEVICE_ID_MATCH_DEVICE
+				| USB_DEVICE_ID_MATCH_INT_INFO,
+	  .idVendor             = 0x045e,
+	  .idProduct            = 0x07bf,
+	  .bInterfaceClass      = USB_CLASS_VIDEO,
+	  .bInterfaceSubClass   = 1,
+	  .bInterfaceProtocol   = 1 },
 	/* Syntek (HP Spartan) */
 	{ .match_flags		= USB_DEVICE_ID_MATCH_DEVICE
 				| USB_DEVICE_ID_MATCH_INT_INFO,
reverted:
--- /var/tmp/portage/sys-kernel/gentoo-sources-4.0.5/work/linux-4.0.5-gentoo/drivers/platform/x86/Kconfig	2015-04-12 17:12:50.000000000 -0500
+++ /usr/src/linux/drivers/platform/x86/Kconfig	2015-06-29 15:36:16.993477459 -0500
@@ -884,4 +884,11 @@
 	  a paravirtualized device provided by QEMU; it lets a virtual machine
 	  (guest) communicate panic events to the host.
 
+config SURFACE_BUTTON
+	tristate "Surface Pro Button Driver"
+	depends on ACPI
+	---help---
+	  This driver provides button support for the surface pro tablet, including
+	  Power Button, Home Button, Volume Button.
+
 endif # X86_PLATFORM_DEVICES
reverted:
--- /var/tmp/portage/sys-kernel/gentoo-sources-4.0.5/work/linux-4.0.5-gentoo/drivers/platform/x86/Makefile	2015-04-12 17:12:50.000000000 -0500
+++ /usr/src/linux/drivers/platform/x86/Makefile	2015-06-29 15:36:16.994477459 -0500
@@ -58,3 +58,4 @@
 
 obj-$(CONFIG_PVPANIC)           += pvpanic.o
 obj-$(CONFIG_ALIENWARE_WMI)	+= alienware-wmi.o
+obj-$(CONFIG_SURFACE_BUTTON)	+= surface_button.o
